airsonic: init at 10.0.0

disassembler opened the pull request:
###### Motivation for this change
addresses issue #28014

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bjornfor commented on nixos/modules/services/misc/airsonic.nix:
```diff
@@ -0,0 +1,119 @@
+  options = {
+
+    services.airsonic = {
+      enable = mkEnableOption "Airsonic";
```
How about "Airsonic, the Free and Open Source media streaming server (fork of Subsonic and Libresonic)"?

disassembler replied:
fixed.

bjornfor commented on nixos/modules/services/misc/airsonic.nix:
```diff
@@ -0,0 +1,119 @@
+
+      listenAddress = mkOption {
+        type = types.string;
+        default = "0.0.0.0";
```
I guess we don't have a policy on this (so do whatever you like), but I think it'd be nice if services defaulted to listen on localhost only.

bjornfor replied:
My reason being that by default you can login with user/password "admin", and the only way to change those credentials is by running the service. So it should not be exposed to the world until it's fully configured.

disassembler replied:
Agreed, security issue here, switched to 127.0.0.1.

bjornfor commented on nixos/modules/services/misc/airsonic.nix:
```diff
@@ -0,0 +1,119 @@
+
+      transcoders = mkOption {
+        type = types.listOf types.path;
+        default = [ "${pkgs.ffmpeg.bin}/bin/ffmpeg" ];
```
I think this needs

```
defaultText = "[ ${pkgs.ffmpeg.bin}/bin/ffmpeg ]";
```

So that the manual can be built without building ffmpeg.

bjornfor replied:
Oops, I guess you're right! (But didn't test.) Grepping for "defaultText" in nixpkgs gives some examples of escaping. I guess try one type of escape (backslash or double single quotes) and check the result?

disassembler replied:
fixed (\ before $)

bjornfor commented on nixos/modules/services/misc/airsonic.nix:
```diff
@@ -0,0 +1,119 @@
+
+  config = mkIf cfg.enable {
+    systemd.services.airsonic = {
+      description = "Personal media streamer";
```
Please capitalize systemd service descriptions. Personal suggestion: "Airsonic Media Server".

disassembler replied:
Done

bjornfor commented on nixos/modules/services/misc/airsonic.nix:
```diff
@@ -0,0 +1,119 @@
+
+      preStart = ''
+        # Install transcoders.
+        ${pkgs.coreutils}/bin/rm -rf ${cfg.home}/transcode ; \
```
You don't need "; \" at the end of each line. preStart is already a shell script.

disassembler replied:
Agreed, this whole module was a copy/paste of subsonic service. I stopped futzing once I got the thing to run. Refactored this whole script. No reason for the bash -c hack either. toString returns a space separated string from an array which is exactly what for needs as an input. I think it had the ridiculous IGNORED_FIRST_ARG because the first argument to a script is the script name itself, which doesn't exist and since it was using $@ it needed something before the actual list.

bjornfor replied:
NixOS has support for directly using upstream units, via the ```systemd.packages``` option. I don't have first-hand experience with using that option (and how one adapts the unit afterwards, with custom options etc.), but you may want to look into it. (Just FYI.)

bjornfor commented on nixos/modules/services/misc/airsonic.nix:
```diff
@@ -0,0 +1,119 @@
+      '';
+      serviceConfig = {
+        # Needed for Airsonic to find airsonic.war.
+        WorkingDirectory = "${pkgs.airsonic}/webapps";
```
"Needed for Airsonic to find airsonic.war"... that refers to the "WorkingDirectory" setting? That's weird... How can it _not_ find the war file that is passed to it on the command line?

disassembler replied:
Agreed, doesn't make sense, pulled it out. Another artifact of the old subsonic.

bjornfor commented on pkgs/servers/misc/airsonic/default.nix:
```diff
@@ -0,0 +1,24 @@
+    platforms = platforms.all;
+    maintainers = with maintainers; [ disassembler ];
+  };
+}
```
Please factor out adding of airsonic package in a separate commit. (That's _commit_ not _PR_.)

disassembler replied:
rebased into two separate commits. Let me know if I need to change the message for the second to something else to meet standards.

bjornfor replied:
The 2nd commit should have prefix "nixos/airsonic:". I suggest "nixos/airsonic: add module".

disassembler replied:
Fixed

bjornfor commented on nixos/modules/services/misc/airsonic.nix:
```diff
@@ -0,0 +1,119 @@
+        ${pkgs.bash}/bin/bash -c ' \
+          for exe in "$@"; do \
+            ${pkgs.coreutils}/bin/ln -sf "$exe" ${cfg.home}/transcode; \
+          done' IGNORED_FIRST_ARG ${toString cfg.transcoders}
```
What's this "IGNORED_FIRST_ARG"?

disassembler replied:
see above comment, no longer applies.

bjornfor reviewed (changes requested):
See above comments.

Commit:
airsonic: init at 10.0.0

Commit:
nixos/airsonic: add module

Mic92 committed:
nixos/airsonic: change script to serviceConfig.ExecStart

- shell invocation is not necessary here

Mic92 committed:
nixos/airsonic: remove full-path commands from preStart

systemd services are initialised with a default PATH.
This path includes coreutils.

Mic92 committed:
airsonic: 10.0.0 -> 10.0.1
